Improvements and commit time rev filter

waltim opened the pull request:
As you requested.

waltim committed:
add new folder to DirectoriesRules, improve git-log commit list filters and close git before return summaries if mainBranch cannot be found

waltim committed:
close walker after collect all files and remove throwing exceptions to cannot break the execution after catch exception

waltim committed:
Refactored classes for improved memory management and performance.

waltim committed:
- Refactored the RepositoryWalker class to eliminate memory leaks and improve performance.
- Replaced the use of synchronized with an ExecutorService for more efficient thread management.
- Reused the same thread pool for multiple metric collection tasks.
- General improvements in code organization and exception handling.
- Optimized parallel metric collection to accelerate the revision analysis process.
- Added explicit disposal of unused objects to free up memory as needed.
- Other minor code improvements and fixes to enhance efficiency and robustness.

These changes significantly enhance the performance and stability of Git project analysis.

rafaelcn commented on src/main/java/br/unb/cic/js/miner/metrics/Summary.java:
```diff
@@ -18,9 +18,7 @@ public class Summary {
 
     public final List<Metric> metrics;
-
-    // Map of files to errors that ocurred when parsing or visiting
```
The comment was just fine, it adds explanation about its usage

rafaelcn commented on src/main/java/br/unb/cic/js/walker/RepositoryWalker.java:
```diff
@@ -71,17 +72,19 @@ public List<Summary> traverse(Date initial, Date end, int steps, int threads) th
             mainBranch = branches.get().getTarget().getName().substring("refs/remotes/origin/".length());
         } else {
             logger.error("{} -- failed to get the project main branch", project);
+            git.close();
```
very important thing to do indeed, I only wonder if we could do it with a `try-with-resources`, although it adds a lot of nested code

rafaelcn commented on src/main/java/br/unb/cic/js/walker/RepositoryWalker.java:
```diff
@@ -233,9 +253,7 @@ private Summary collect(ObjectId head, Date current, Map<Date, ObjectId> commits
-            ex.printStackTrace();
-
-            throw new RuntimeException(ex);
+            System.out.println(ex.getStackTrace());
```
That's exactly the same thing as `ex.printStackTrace()`, isn't it?

rafaelcn commented on src/main/java/br/unb/cic/js/walker/RepositoryWalkerTask.java:
```diff
@@ -12,51 +12,47 @@
-
     public final Interval interval;
-
-    // Threads represents the number of visitors working concurrently on a given revision
```
Important description of the wording here as we use threads in multiple places with different meaning, as well as the comment about the output variable

rafaelcn commented on src/main/java/br/unb/cic/js/walker/RepositoryWalkerTask.java:
```diff
@@ -12,51 +12,47 @@
-            summaries.forEach(s -> {
-                        content.append(s.values())
-                       .append("\n");
+            ConcurrentLinkedQueue<Summary> summaries = new ConcurrentLinkedQueue<>();
```
That's the same thing as `val summaries = walker.traverse()`, care to give me your rational about splitting definition and attribution? looks like code repetition to me

rafaelcn replied:
Another thing I'd mention is that you don't actually need a ConcurrentLinkedQueue over here since the summaries are already complete and that's the only step running concurrently when the projects threads are set to 1.

rafaelcn commented on src/main/java/br/unb/cic/js/walker/RepositoryWalkerTask.java:
```diff
@@ -69,19 +65,14 @@ public void run() {
-            throw new RuntimeException(ex);
         } catch (Exception ex) {
+            ex.printStackTrace();
+            System.out.println(ex.getMessage());
```
don't use `System.out.println`, use the logger with a replacement variable such as the one in line 83.

Example: `logger.error("failed to traverse project {}, reason {}", walker.project, ex.getMessage());`

rafaelcn commented on src/main/java/br/unb/cic/js/walker/Walker.java:
```diff
@@ -139,10 +139,11 @@ public void traverse() {
         } catch (java.lang.InterruptedException | java.util.concurrent.ExecutionException ex) {
             logger.error("failed to execute a concurrent task, reason {}", ex.getMessage());
-            ex.printStackTrace();
+            // ex.printStackTrace();
```
The stack trace is important in both cases, leave them as the message may not be the best thing when debugging an exception

rafaelcn reviewed (changes requested):
I made some questions and raised some concerns about the writing of the code

waltim committed:
Optimized RepositoryWalker.collect method for better performance and fixed recursion issues.

waltim committed:
Refactored the RepositoryWalkerTask class for improved performance and code readability.
Introduced proper handling of metrics to ensure only non-null summaries are processed.
Fixed issues related to writing metrics to CSV files with proper line breaks.
Enhanced file filtering logic for collecting JavaScript files.
Removed unnecessary validation logic and improved metric completeness check.
Updated file collection logic to avoid circular references.
Reorganized the placement of metrics calculation within the code.
Improved thread management and executor shutdown logic.
Addressed issues related to capturing metrics from JavaScript files.
Ensured complete metric data is stored in CSV files as expected.
These changes enhance the performance, readability, and reliability of the code for the RepositoryWalkerTask class, improving the overall functionality of the program.

waltim committed:
Refactored the RepositoryWalkerTask class for improved performance and code readability.
Introduced proper handling of metrics to ensure only non-null summaries are processed.
Fixed issues related to writing metrics to CSV files with proper line breaks.
Enhanced file filtering logic for collecting JavaScript files.
Removed unnecessary validation logic and improved metric completeness check.
Updated file collection logic to avoid circular references.
Reorganized the placement of metrics calculation within the code.
Improved thread management and executor shutdown logic.
Addressed issues related to capturing metrics from JavaScript files.
Ensured complete metric data is stored in CSV files as expected.
